src/dex/balancer-v2/LinearPool.ts

```diff
@@ -0,0 +1,372 @@
+    i.e. 
+    scalingFactors: pool.tokens.map(({ decimals, priceRate }) =>
+        MathSol.mulDownFixed(getTokenScalingFactor(decimals), priceRate)
+    )
+    */
+  onSell(amounts: bigint[], poolPairData: LinearPoolPairData): bigint[] {
+    return this._swapGivenIn(
+      amounts,
+      poolPairData.tokens,
+      poolPairData.balances,
+      poolPairData.indexIn,
+      poolPairData.indexOut,
+      poolPairData.bptIndex,
+      poolPairData.wrappedIndex,
+      poolPairData.mainIndex,
+      poolPairData.scalingFactors,
+      poolPairData.swapFee,
+      poolPairData.lowerTarget,
+      poolPairData.upperTarget,
+    );
+  }
+
+  _swapGivenIn(
+    tokenAmountsIn: bigint[],
+    tokens: string[],
+    balances: bigint[],
+    indexIn: number,
+    indexOut: number,
+    bptIndex: number,
+    wrappedIndex: number,
+    mainIndex: number,
+    scalingFactors: bigint[],
+    swapFeePercentage: bigint,
+    lowerTarget: bigint,
+    upperTarget: bigint,
+  ): bigint[] {
+    /* 
+        Linear pools allow trading between:
+        wrappedToken <> mainToken
+        wrappedToken <> BPT
+        mainToken <> BPT
+        */
+    let pairType: PairTypes;
+    if (isSameAddress(tokens[indexIn], tokens[bptIndex])) {
+      if (isSameAddress(tokens[indexOut], tokens[wrappedIndex]))
+        pairType = PairTypes.BptToWrappedToken;
+      else pairType = PairTypes.BptToMainToken;
+    } else if (isSameAddress(tokens[indexOut], tokens[bptIndex])) {
+      if (isSameAddress(tokens[indexIn], tokens[wrappedIndex]))
+        pairType = PairTypes.WrappedTokenToBpt;
+      else pairType = PairTypes.MainTokenToBpt;
+    } else {
+      if (isSameAddress(tokens[indexIn], tokens[wrappedIndex]))
+        pairType = PairTypes.WrappedTokenToMainToken;
+      else pairType = PairTypes.MainTokenToWrappedToken;
+    }
+
+    const balancesUpscaled = this._upscaleArray(balances, scalingFactors);
+    const tokenAmountsInScaled = tokenAmountsIn.map(a =>
+      this._upscale(a, scalingFactors[indexIn]),
+    );
+
+    // VirtualBPTSupply must be used for the maths
+    const virtualBptSupply = this.MAX_TOKEN_BALANCE.sub(
+      balances[bptIndex],
+    ).toBigInt();
+
+    const amountsOut = this._onSwapGivenIn(
+      tokenAmountsInScaled,
+      balancesUpscaled[mainIndex],
+      balancesUpscaled[wrappedIndex],
+      swapFeePercentage,
+      lowerTarget,
+      upperTarget,
+      virtualBptSupply,
+      pairType,
+    );
+
+    // amountOut tokens are exiting the Pool, so we round down.
+    return amountsOut.map(a =>
+      this._downscaleDown(a, scalingFactors[indexOut]),
+    );
+  }
+
+  /*
+     Called when a swap with the Pool occurs, where the amount of tokens entering the Pool is known.
+     All amounts are upscaled.
+     Swap fee is NOT already deducted.
+     The return value is also considered upscaled, and should be downscaled (rounding down)
+     */
+  _onSwapGivenIn(
```

Comment:
Is this function complete? I am not sure if all the cases are handled as per the contract https://github.com/balancer-labs/balancer-v2-monorepo/blob/master/pkg/pool-linear/contracts/LinearPool.sol#L253

Reply:
We have 6 pair types but we are only handling 2?